crates/holochain_state/src/source_chain.rs

```diff
@@ -555,8 +562,22 @@ where
         Ok(self.scratch.apply(|scratch| scratch.records().collect())?)
     }
 
-    pub fn has_initialized(&self) -> SourceChainResult<bool> {
-        Ok(self.len()? > 3)
+    pub async fn zomes_initialized(&self) -> SourceChainResult<bool> {
+        if self.zomes_initialized.load(Ordering::Relaxed) {
+            return Ok(true);
+        }
+        let query_filter = ChainQueryFilter {
+            action_type: Some(ActionType::InitZomesComplete),
+            ..QueryFilter::default()
+        };
+        let init_zomes_complete_actions = self.query(query_filter).await?;
+        let zomes_initialized = init_zomes_complete_actions.len() == 1;
```

Comment:
We could change this to `len() > 0`, but I think it must never happen that it's greater than 1. If it's > 1 the init zomes check is wrong.

Reply:
I agree. Even though it should never be `> 1`, it's more correct to accept that than to return the wrong answer, and the presence of multiple causes no harm.

```suggestion
        if init_zomes_complete_actions.len() > 1 {
            tracing::warn("Multiple InitZomesComplete actions are present");
        }
        let zomes_initialized = init_zomes_complete_actions.len() > 0;
```